### DISCUSSION TITLE
Notify Slack for the first success of a job after failures

### DISCUSSION THREAD

**Discussion by OddBloke** [concourse/concourse#7919] [Help & Support]
Hey folks,

We have a Concourse job which runs every 10m to perform a batch operation (in this case, pushing metrics into StatsD). Getting a Slack message every time this runs is way too noisy, so we have Concourse configured to notify Slack on abort/error/failure only. However, this means that if you're looking at Slack, you might think that the job only fails: you have to click through to the web UI to see if there have been successful builds since the last failure.

I would like to be able to configure a notification hook which runs only when a job has moved from {abort,error,failure} to success: we would therefore receive notifications for failures and for recoveries from failures, improving the experience from a Slack POV. is there any way to do this?

Thanks!

Dan

**Comment by suhlig** [comment]
I had a similar requirement and solved it with an external counter (external state, basically). As far as I know in Concourse, any persistent state beyond the lifetime of a job needs to be represented externally.

In my current solution a failed task has an `on_error` step that increases a number in a file and pushes it to GitHub. When the task succeeds (`on_success`) the counter is reset to zero. Another task looks at the counter and alerts if it exceeds a threshold or moves to zero.

I also thought of a better way: The counter could be represented in Redis, which has counters as a first-class concept. But I did not find an existing resource and also haven't gotten around to write one yet.

**Comment by marco-m** [comment]
I was tempted to answer like suhlig above, that you need to setup yourself an external storage for status, but then, thinking how to do it with a resource as opposed to a task, I realized the following.
It should also be possible to do it from within the notification resource: on put, the resource gets "metadata" (https://concourse-ci.org/implementing-resource-types.html#resource-metadata), among which there is enough information to know the job number.
The resource could query the ATC (same query as fly) and find the status of the previous run. Based on that, it could decide wether to notify or not.

**Reply by suhlig** [reply]
I like that idea, but how would you get the credentials needed for `fly`? The beauty of [`set_pipeline`](https://concourse-ci.org/set-pipeline-step.html) is that you don't need credentials, but it's a built-in. We'd need something similar for self-modifying pipelines and the scenario of this thread.
